### PR TITLE
feat(core): avatar group horizon

### LINKED ISSUE #9826: Component Design Update: Avatar Group
- [ ] review markup changes from fd-styles
- [ ] Horizon specs(visual and interaction) updates
- [ ] Delta theming
- [ ] identify and document breaking changes

### PR DESCRIPTION
**Pull request by g-cheishvili**
## Related Issue(s)

closes #9826

## Description

1. Implemented a new Avatar group which works with new fundamental-styles
2. Moved old avatar group to `legacy` library

### CODE REVIEW

File: libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        '[class.fd-avatar-group--xs]': 'size === "xs"',
+        '[class.fd-avatar-group--s]': 'size === "s"',
+        '[class.fd-avatar-group--m]': 'size === "m"',
+        '[class.fd-avatar-group--l]': 'size === "l"',
+        '[class.fd-avatar-group--xl]': 'size === "xl"'
+    },
+    template: '<ng-content></ng-content>',
+    standalone: true,
+    imports: [NgIf],
+    providers: [DestroyedService],
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts:39]: Replace with DestroyRef?

File: libs/core/src/lib/avatar-group/components/avatar-group-overflow-body/avatar-group-overflow-body.component.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+/**
+ * Avatar group overflow body component, used to provide a template for the avatar group overflow body.
+ */
+@Component({
+    selector: 'fd-avatar-group-overflow-body',
+    templateUrl: './avatar-group-overflow-body.component.html',
+    host: {
+        class: 'fd-popover__wrapper',
+        '[style.display]': '"flex"',
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/components/avatar-group-overflow-body/avatar-group-overflow-body.component.ts:11]: Is there a season this style cannot be added to .scss file?

File: libs/core/src/lib/avatar-group/components/avatar-group-overflow-button.component.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+ * Avatar group overflow button component.
+ * This is used for the overflow button template.
+ * Although any content can be projected as the overflow button, this component provides some styling and accessibility features.
+ */
+@Component({
+    selector: 'fd-avatar-group-overflow-button',
+    template: ` <ng-content></ng-content>`,
+    host: {
+        role: 'button',
+        '[class.fd-avatar--circle]': 'circle'
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/components/avatar-group-overflow-button.component.ts:23]: Use HostBinding decorator instead

File: libs/core/src/lib/avatar-group/components/default-avatar-group-overflow-body/default-avatar-group-overflow-body.component.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        PopoverBodyHeaderDirective,
+        BarModule
+    ],
+    host: {
+        class: 'fd-popover__wrapper',
+        '[style.display]': '"flex"',
+        '[style.flex-direction]': '"column"',
+        '[style.max-width.rem]': '20'
+    },
+    standalone: true
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/components/default-avatar-group-overflow-body/default-avatar-group-overflow-body.component.ts:41]: Change Detection, View encapsulation

File: libs/core/src/lib/avatar-group/components/avatar-group-overflow-button.component.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    },
+    hostDirectives: [NgClass],
+    changeDetection: ChangeDetectionStrategy.OnPush,
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/components/avatar-group-overflow-button.component.ts:26]: View Encapsulation

File: libs/core/src/lib/avatar-group/components/avatar-group-overflow-body/avatar-group-overflow-body.component.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        '[style.flex-direction]': '"column"'
+    },
+    standalone: true
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/components/avatar-group-overflow-body/avatar-group-overflow-body.component.ts:14]: Change Detection, View encapsulation

File: libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    changeDetection: ChangeDetectionStrategy.OnPush,
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts:40]: View Encapsulation

File: libs/core/src/lib/avatar-group/components/default-avatar-group-overflow-body/default-avatar-group-overflow-body.component.ts [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        AvatarGroupItemRendererDirective,
+        FocusableListDirective,
+        NgIf,
+        NgTemplateOutlet,
+        PopoverBodyHeaderDirective,
+        BarModule
+    ],
+    host: {
+        class: 'fd-popover__wrapper',
+        '[style.display]': '"flex"',
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/components/default-avatar-group-overflow-body/default-avatar-group-overflow-body.component.ts:37]: Same question about moving inline styling into .scss file

File: libs/core/src/lib/avatar-group/project.json [hunk 9]
```diff
@@ -1,32 +1,26 @@
...
-    "sourceRoot": "libs/core/src/lib/avatar-group",
+    "sourceRoot": "libs/core/src/lib/avatar-group/src",
     "prefix": "fd",
+    "tags": ["scope:fd", "type:lib"],
+    "projectType": "library",
     "targets": {
         "build": {
-            "executor": "@nx/angular:ng-packagr-lite",
-            "outputs": ["{workspaceRoot}/dist/libs/core/avatar-group"],
+            "executor": "@nx/angular:package",
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/project.json:10]: Is there a reason executor being changed?

File: libs/core/src/lib/avatar-group/tsconfig.spec.json [hunk 10]
```diff
@@ -6,6 +6,6 @@
         "target": "es2016",
         "types": ["jest", "node"]
     },
-    "files": ["src/test-setup.ts"],
-    "include": ["jest.config.ts", "**/*.test.ts", "**/*.spec.ts", "**/*.d.ts"]
+    "files": ["./test-setup.ts"],
+    "include": ["jest.config.ts", "src/**/*.test.ts", "src/**/*.spec.ts", "src/**/*.d.ts"]
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/tsconfig.spec.json:10]: fix paths to .spec files

File: libs/core/src/lib/avatar-group/tsconfig.json [hunk 11]
```diff
@@ -1,23 +1,17 @@
...
+        "target": "es2022",
+        "useDefineForClassFields": false
+    },
     "files": [],
     "include": [],
     "references": [
         {
             "path": "./tsconfig.lib.json"
         },
-        {
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/tsconfig.json:9]: Was it intentional?

**Review by N1XUS — CHANGES_REQUESTED** [review]

File: libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts (hunk 1 above)
Comment (g-cheishvili) [reply libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts:39]: Thanks, forgot about it during rebase

File: libs/core/src/lib/avatar-group/project.json (hunk 9 above)
Comment (g-cheishvili) [reply libs/core/src/lib/avatar-group/project.json:10]: Not really, just the nx lib generator incorrect config

File: libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    /** @hidden */
+    @Input()
+    items: QueryList<AvatarGroupItemDirective>;
+
+    /** @hidden */
+    @ContentChildren(AvatarGroupItemRendererDirective, { descendants: true })
+    portals: QueryList<AvatarGroupItemRendererDirective>;
+
+    /** @hidden */
+    elementRef = inject(ElementRef);
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts:72]: for all `hidden` items we use `_` prefix

File: libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+    /** @hidden */
+    hiddenItems = signal<AvatarGroupItemRendererDirective[]>([]);
+
+    /** @hidden */
+    private _destroyRef = inject(DestroyRef);
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts:78]: an be also `readonly`

File: libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+        const hiddenItems: AvatarGroupItemRendererDirective[] = [];
+        let accWidth = items.reduce((acc, item) => (item.forceVisibility ? acc + item.width : acc), 0);
+        for (const item of items) {
+            accWidth += item.width;
+            if (accWidth <= containerWidth) {
+                visibleItems.push(item);
+            } else if (!item.forceVisibility) {
+                hiddenItems.push(item);
+            }
+        }
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts:138]: Here we do multiple iterations over one array. Also in subscribe callback another iteration over visible and hidden items. Can this be combined somehow? Maybe in `for` loop you can call `show` or `hide` methods?

Comment (N1XUS) [reply libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts:138]: Another thing is that for `accWidth` you iterate over all items, even though you are interested only in items that were already filtered for the `visibleItems` variable.

File: libs/core/src/lib/avatar-group/components/default-avatar-group-overflow-body/default-avatar-group-overflow-body.component.ts [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+    private changeDetectorRef = inject(ChangeDetectorRef);
+
+    /** @hidden */
+    private _itemClickSubscription: Subscription;
+
+    /** @hidden */
+    private _renderer = inject(Renderer2);
+
+    /** @hidden */
+    private _rtlService = inject(RtlService, { optional: true });
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/components/default-avatar-group-overflow-body/default-avatar-group-overflow-body.component.ts:84]: readonly

File: libs/core/src/lib/avatar-group/directives/avatar-group-internal-overflow-button.directive.ts [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+        const hiddenItems = value || [];
+        if (hiddenItems.length > 0) {
+            this.show(hiddenItems);
+        } else {
+            this.hide();
+        }
+    }
+
+    /** @hidden */
+    readonly templateRef: TemplateRef<{ $implicit: AvatarGroupItemRendererDirective[] }> = inject(TemplateRef);
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/directives/avatar-group-internal-overflow-button.directive.ts:21]: You can also provide `static ngTemplateContextGuard` for type support in templates

**Review by N1XUS — CHANGES_REQUESTED** [review]

File: libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts (hunk 12 above)
Comment (g-cheishvili) [reply libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts:72]: This comes from the interface, better would be to add repetitive jsdoc just to align with the commonness

File: libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts (hunk 14 above)
Comment (g-cheishvili) [reply libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts]: It's down to 2n now. I'd prefer to keep the subscribe code as-is

Comment (g-cheishvili) [reply libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts]: I've thought about it more and

I need initially visibleItems to contain all the forceVisibleItems and to know how much is their width. If I start iterating and accumulate items, and overflow the allowed width and at the same time I have forceVisibility items left in the queue, I end up in the situation where I will have to iterate back and remove previously added items. I could not think of any better approach :/

File: libs/cdk/src/lib/utils/directives/resize-observer/resize-observer.directive.ts [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+    private _cdr = inject(ResizeObserverChangeDetectorRef, { optional: true }) || inject(ChangeDetectorRef);
+
+    /** @hidden
+     * Since resize observer does not invoke change detection, we need to do it manually.
+     * */
+    // eslint-disable-next-line @typescript-eslint/member-ordering
+    @Output()
+    resized: Observable<ResizeObserverEntry[]> = this._resizeObserverService
+        .observe(this._elementRef.nativeElement)
+        .pipe(tap(() => this._cdr.detectChanges()));
```
Comment (N1XUS) [inline libs/cdk/src/lib/utils/directives/resize-observer/resize-observer.directive.ts:31]: Should it really detect changes? I think the consumer should decide that and not the "provider"

File: libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+        class: 'fd-avatar-group',
+        '[class.fd-avatar-group--individual-type]': 'type === "individual"',
+        '[class.fd-avatar-group--group-type]': 'type === "group"',
+        '[class.fd-avatar-group--horizontal]': 'orientation === "horizontal"',
+        '[class.fd-avatar-group--vertical]': 'orientation === "vertical"',
+        '[class.fd-avatar-group--xs]': 'size === "xs"',
+        '[class.fd-avatar-group--s]': 'size === "s"',
+        '[class.fd-avatar-group--m]': 'size === "m"',
+        '[class.fd-avatar-group--l]': 'size === "l"',
+        '[class.fd-avatar-group--xl]': 'size === "xl"'
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/components/avatar-group-host.component.ts:37]: In `AvatarGroupOverflowButtonComponent` you use NgClass directive to compose complex classList, here you use host. Any particular reason to use different style?

File: libs/core/src/lib/avatar-group/components/avatar-group-overflow-button.component.ts [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+    @HostBinding('class.fd-avatar--circle')
+    circle = false;
+
+    /**
+     * A number from 1 to 10 representing the background color of the Avatar.
+     */
+    @Input() colorAccent: Nullable<ColorAccent> = null;
+
+    /** @hidden */
+    private ngClassDirective = inject(NgClass);
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/components/avatar-group-overflow-button.component.ts:64]: `_` prefix

File: libs/core/src/lib/avatar-group/components/avatar-group-overflow-button.component.ts [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+        this.updateClass();
+    }
+
+    /** @hidden */
+    ngOnChanges(): void {
+        this.updateClass();
+    }
+
+    /** @hidden */
+    private updateClass(): void {
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/components/avatar-group-overflow-button.component.ts:83]: `_` prefix

File: libs/core/src/lib/avatar-group/components/default-avatar-group-overflow-body/default-avatar-group-overflow-body.component.html [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+</div>
+<div
+    class="fd-avatar-group__overflow-body"
+    [class.fd-avatar-group__overflow-body--no-padding]="_isDetailStage"
+    [class.fd-avatar-group__overflow-body--no-horizontal-scroll]="_isDetailStage"
+    fdkFocusableList
+    navigationDirection="horizontal"
+>
+    <ng-container *ngIf="_overflowPopoverStage === 'main'; else detailsStage">
+        <ng-container *ngFor="let item of avatars">
```
Comment (N1XUS) [inline libs/core/src/lib/avatar-group/components/default-avatar-group-overflow-body/default-avatar-group-overflow-body.component.html:27]: Maybe worth considering adding trackBy function?

**Review by N1XUS — APPROVED** [review]

**Pull request merged**